llvm/test/MC/ARM/pcrel-arm-ldr-imm8-relocs.s

```diff
@@ -0,0 +1,41 @@
+@ RUN: llvm-mc -filetype=obj -triple=armv7 %s -o %t
+@ RUN: llvm-readelf -r %t | FileCheck %s --check-prefix=ARM
+@ RUN: llvm-objdump -d --triple=armv7 %t | FileCheck %s --check-prefix=ARM_ADDEND
+
+@ ARM: R_ARM_LDRS_PC_G0
+@ ARM: foo1
+@ ARM: R_ARM_LDRS_PC_G0
+@ ARM: foo2
+@ ARM: R_ARM_LDRS_PC_G0
+@ ARM: foo3
+
+// The value format is decimal in these specific cases, but it's hex for other
+// ldr instructions. These checks are valid for both formats.
+
+@ ARM_ADDEND: r0, [pc, #-
```

Comment:
Are these offsets the way they are (all -8) because they come from unresolved locations? So later, that -8 becomes the proper offset?

If not you'd expect references to foo1, foo2 and foo3 to have different immediates here.

But I think it's the first thing, and I have just forgotten how relocations work :)

Reply:
Istr something about the value of the PC while executing an instruction at address N is actually N+something (8?) so these would be loading from themselves. But only because they're waiting for a linker to fill in the real value.

Reply:
The -8 is the addend for arm pcrel instructions because of the PC offset. Technically the default offset (=the immediate value in the instruction before the relocation is applied) should be 0 to indicate the current PC, but the PC is 2 instructions behind for historical reasons. So for Thumb the added is -4. It's mentioned it in the ABI -[ ](https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst#addends-and-pc-bias-compensation)

We can't see the final value here because the relocation resolution is a linker's job and the linker is not run.

Reply:
Got it, thanks for explaining.

Reply:
I was wrong actually. The PC is 2 instructions ahead, hence the the -8 addend offsets it to the right instruction ;)